### PR TITLE
machine/rp2040: implement Flash interface

### PR DESCRIPTION
**Pull request by deadprogram**
This PR does for the RP2040 what we've got for the NRF, SAMD21, etc.

```
$ tinygo flash -target pico -monitor examples/flash
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
Flash data start:       0x10003000
Flash data end:         0x10200000
Flash data size, bytes: 2084864
Flash write block size: 4
Flash erase block size: 4096

Reading data from flash: 1234567887654321123456788765432112345678876543211234567887654321
Writing data to flash: 1234567887654321123456788765432112345678876543211234567887654321
Reading data back from flash: 1234567887654321123456788765432112345678876543211234567887654321
```

This branch has both the @kenbell #3382 and a heavily modified version of my own #3378 but of course, actually working!

### CODE REVIEW

File: src/machine/machine_rp2040_rom.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+	return len(p), nil
+}
+
+// WriteAt writes the given number of bytes to the block device.
+// Only word (32 bits) length data can be programmed.
+// If the length of p is not long enough it will be padded with 0xFF bytes.
+// This method assumes that the destination is already erased.
+//
+//go:section .ramtext
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go]: The section name is out of date.  If it's working, then probably don't need `go:section` here?

File: src/machine/machine_rp2040_rom.go [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+func (f flashBlockDevice) WriteAt(p []byte, off int64) (n int, err error) {
+	if uintptr(off)+uintptr(len(p)) > FlashDataEnd() {
+		return 0, errFlashCannotWritePastEOF
+	}
+
+	// state := interrupt.Disable()
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go]: Remove commented code?

File: src/machine/machine_rp2040_rom.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	// 0x00003000
+	address := uintptr(off) - C.XIP_BASE
+	padded := f.pad(p)
+
+	writeFlash(padded, address)
+
+	return len(padded), nil
+}
+
+//go:section .ramtext
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go]: as above fix section name, or remove if not needed

File: src/machine/machine_rp2040_rom.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+// EraseBlockSize to map addresses to blocks.
+func (f flashBlockDevice) EraseBlocks(start, length int64) error {
+	// state := interrupt.Disable()
+	// defer interrupt.Restore(state)
+
+	address := uintptr(start*f.EraseBlockSize()) - C.XIP_BASE
+	return eraseFlashBlocks(address, length*f.EraseBlockSize())
+}
+
+//go:section .ramtext
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go]: ditto

File: src/machine/machine_rp2040_rom.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
+//go:build rp2040
+
+package machine
+
+import (
+	"bytes"
+	//"runtime/interrupt"
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go]: Remove commented code?

**Review by kenbell — CHANGES_REQUESTED** [review]
A few nits

File: src/machine/machine_rp2040_rom.go (hunk 1 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go]: Removed, thanks for pointing that out.

File: src/machine/machine_rp2040_rom.go (hunk 2 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go]: Actually I re-enabled it.

File: src/machine/machine_rp2040_rom.go (hunk 3 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go]: :scissors:

File: src/machine/machine_rp2040_rom.go (hunk 4 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go]: natch

File: src/machine/machine_rp2040_rom.go (hunk 5 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go]: re-activated

**Comment by deadprogram** [comment]
Thanks for the feedback @kenbell I was able to remove a bunch of extraneous code.

**Comment by bgould** [comment]
seeing panic testing with Feather RP2040: https://github.com/tinygo-org/tinyfs/blob/machine-flash/examples/console/littlefs/machine/main.go

```
~/src/tinyfs/examples/console/littlefs/machine (machine-flash)

--- Miniterm on COM16  9600,8,N,1 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---

==> lsblk

-------------------------------------
 Device Information:
-------------------------------------
 flash data start: 10016000
 flash data end:  10800000
-------------------------------------

==> format
panic: runtime error: unsafe.Slice/String: len out of range
```

**Comment by bgould** [comment]
Same result testing with Pico W

**Comment by deadprogram** [comment]
@bgould I will look into it by the weekend. Thanks for testing.

**Comment by deadprogram** [comment]
```
$ tinygo flash -target pico -monitor ./examples/console/littlefs/machine/
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
SPI Configured. Reading flash info
==> lsblk

-------------------------------------
 Device Information:  
-------------------------------------
 flash data start: 10016000
 flash data end:  10200000
-------------------------------------

==> format
go_lfs_block_device_read: 0x20002098, 1, 0, 0x20002328, 4
go_lfs_block_device_erase: 0x20002098, 0
go_lfs_block_device_read: 0x20002098, 0, 52, 0x20002328, 4
go_lfs_block_device_prog: 0x20002098, 0, 0, 0x20002528, 52
go_lfs_block_device_sync: 0x20002098
go_lfs_block_device_read: 0x20002098, 0, 0, 0x20002328, 52
panic: runtime error: unsafe.Slice/String: len out of range
```

`0x20002098` looks like a rather invalid address to me!

**Comment by deadprogram** [comment]
I was misreading that debug output. Seems like it is referring to block 0 offset 0.

**Comment by kenbell** [comment]
@deadprogram  I may be missing something - but is the RP2040 implementation mixing up offset and absolute addresses vs the nrf implementation?

From nrf:
```
func (f flashBlockDevice) EraseBlocks(start, len int64) error {
	address := FlashDataStart() + uintptr(start*f.EraseBlockSize())
```

From RP2040:
```
func (f flashBlockDevice) EraseBlocks(start, length int64) error {
	state := interrupt.Disable()
	defer interrupt.Restore(state)

	address := uintptr(start*f.EraseBlockSize()) - C.XIP_BASE
```

It looks like the RP2040 implementation assumes start is an absolute memory address whereas nrf implementation assumes it's an offset from FlashDataStart.

**Comment by deadprogram** [comment]
I was just dealing with that exact realization! :) Making some changes to the other platforms like STM32 as a result.

**Comment by deadprogram** [comment]
IN short, I am making it consistent that `start` will always mean "relative to `FlashDataStart()` and make adjustments accordingly.

**Comment by deadprogram** [comment]
Made suitable corrections, thanks for your help @bgould and @kenbell 

```
$ tinygo flash -target pico -monitor examples/flash
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
Flash data start:       0x10003000
Flash data end:         0x10200000
Flash data size, bytes: 2084864
Flash write block size: 4
Flash erase block size: 4096

Reading data from flash: 1234567887654321123456788765432112345678876543211234567887654321
Writing data to flash: 1234567887654321123456788765432112345678876543211234567887654321
Reading data back from flash: 1234567887654321123456788765432112345678876543211234567887654321
```

also

```
$ tinygo flash -target pico -monitor ./examples/console/littlefs/machine/
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
SPI Configured. Reading flash info
==> lsblk

-------------------------------------
 Device Information:  
-------------------------------------
 flash data start: 10016000
 flash data end:  10200000
-------------------------------------

==> format
Successfully formatted LittleFS filesystem.
```

I will need to submit an additional PR with some fixes to STM32 that I discovered as part of doing this PR.

**Comment by mytechnotalent** [comment]
@deadprogram I would like to volunteer to test this once you are complete if you need help as persistent storage for the RP2040 is critical.

File: src/machine/machine_rp2040_rom.go [hunk 6]
```diff
@@ -145,7 +145,8 @@ func EnterBootloader() {
 }
 
 // Flash related code
-const memoryStart = 0
+// const memoryStart = 0 // memory start for purpose of erase
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go]: Remove this line?

File: src/machine/machine_rp2040_rom.go [hunk 7]
```diff
@@ -242,3 +243,13 @@ func (f flashBlockDevice) pad(p []byte) []byte {
 	padding := bytes.Repeat([]byte{0xff}, int(paddingNeeded))
 	return append(p, padding...)
 }
+
+// return the correct address to be used for write
+func writeAddress(off int64) uintptr {
+	return FlashDataStart() - uintptr(C.XIP_BASE) + uintptr(off)
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go]: Relies on correct understanding of operator precedence, perhaps re-order as `return FlashDataStart() + uintptr(off) - uintptr(C.XIP_BASE)` or add brackets?

File: src/machine/flash.go [hunk 8]
```diff
@@ -91,21 +95,22 @@ func (fl *FlashBuffer) Write(p []byte) (n int, err error) {
...
-		fl.current += uintptr(n)
-		return n, nil
+
+	// calculate currentPageBlock relative to fl.start, meaning that
+	// block 0 -> fl.start
+	// block 1 -> fl.start + pagesize
+	// block 2 -> fl.start + pagesize*2
+	// ...
+	currentPageBlock := (fl.start + fl.offset - FlashDataStart()) / pagesize
+	lastPageBlockNeeded := (fl.start + fl.offset + uintptr(len(p)) - FlashDataStart()) / pagesize
```
Comment (kenbell) [inline src/machine/flash.go]: This no longer works for partial page writes, old logic contained `+ pagesize - 1` to allow for partial page writes.

Does it need documenting that len(p) must be a multiple of `EraseBlockSize()`?   (or restore the old logic)

File: src/machine/flash.go [hunk 9]
```diff
@@ -91,21 +95,22 @@ func (fl *FlashBuffer) Write(p []byte) (n int, err error) {
...
+
+	// so we create at least 1 page
+	if currentPageBlock == lastPageBlockNeeded {
```
Comment (kenbell) [inline src/machine/flash.go]: Is this the correct logic?  If I do a zero-byte write, should the function erase the block, or just return as a no-op?

Is this compensating for the change above that no longer supports partial page writes?

**Review by kenbell — COMMENTED** [review]
Some more minor nits

File: src/machine/flash.go (hunk 8 above)
Comment (deadprogram) [reply src/machine/flash.go]: You are correct, I am modifying my code to restore that capability.

File: src/machine/flash.go (hunk 9 above)
Comment (deadprogram) [reply src/machine/flash.go]: Yes, so I can remove this by restoring the partial page write as you suggest.

Comment (deadprogram) [reply src/machine/flash.go]: Done.

File: src/machine/flash.go (hunk 8 above)
Comment (deadprogram) [reply src/machine/flash.go]: Done.

File: src/machine/machine_rp2040_rom.go (hunk 7 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go]: Done.

File: src/machine/machine_rp2040_rom.go (hunk 6 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go]: Done.

File: src/machine/flash.go [hunk 10]
```diff
@@ -114,11 +114,11 @@ func (fl *FlashBuffer) Write(p []byte) (n int, err error) {
 			last = len(p)
 		}
 
-		_, err := fl.b.WriteAt(p[i:last], int64(fl.current))
+		_, err := fl.b.WriteAt(p[i:last], int64(fl.offset))
 		if err != nil {
 			return 0, err
 		}
-		fl.current += uintptr(last - i)
+		fl.offset += uintptr(i)
```
Comment (kenbell) [inline src/machine/flash.go]: Should this be `fl.offset += pagesize`?    On the 3rd page of a multi-page write, `i` will be `3*pagesize`, so incrementing by that seems bad...

File: src/machine/flash.go [hunk 11]
```diff
@@ -133,12 +133,7 @@ func (fl *FlashBuffer) Close() error {
 // You can only seek relative to the start.
 // Also, you cannot use seek before write operations, only read.
 func (fl *FlashBuffer) Seek(offset int64, whence int) (int64, error) {
```
Comment (kenbell) [inline src/machine/flash.go:137]: This only implements SEEK_START - should at least panic for SEEK_OFFSET and SEEK_END?

File: src/machine/flash.go [hunk 12]
```diff
@@ -66,21 +66,25 @@ type BlockDevice interface {
...
 // OpenFlashBuffer opens a FlashBuffer.
 func OpenFlashBuffer(b BlockDevice, address uintptr) *FlashBuffer {
-	return &FlashBuffer{b: b, start: address, current: address}
+	return &FlashBuffer{b: b, start: address}
 }
 
 // Read data from a FlashBuffer.
 func (fl *FlashBuffer) Read(p []byte) (n int, err error) {
-	fl.b.ReadAt(p, int64(fl.current))
+	fl.b.ReadAt(p, int64(fl.offset))
```
Comment (kenbell) [inline src/machine/flash.go]: This just swallows errors, should check for error and reflect it out?

File: src/machine/machine_rp2040_rom.go [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+// in bytes. This is used for the block size in EraseBlocks.
+func (f flashBlockDevice) EraseBlockSize() int64 {
+	return eraseBlockSize()
+}
+
+// EraseBlocks erases the given number of blocks. An implementation may
+// transparently coalesce ranges of blocks into larger bundles if the chip
+// supports this. The start and len parameters are in block numbers, use
+// EraseBlockSize to map addresses to blocks.
+func (f flashBlockDevice) EraseBlocks(start, length int64) error {
```
Comment (kenbell) [inline src/machine/machine_rp2040_rom.go:223]: Read and Write do range checks, but this function doesn't?

**Review by kenbell — COMMENTED** [review]
Few more comments.  Sorry, should have found a couple of these before.

File: src/machine/flash.go (hunk 10 above)
Comment (deadprogram) [reply src/machine/flash.go]: Updated to handle this.

File: src/machine/flash.go (hunk 11 above)
Comment (deadprogram) [reply src/machine/flash.go:137]: Done.

File: src/machine/flash.go (hunk 12 above)
Comment (deadprogram) [reply src/machine/flash.go]: Good point. Done.

File: src/machine/machine_rp2040_rom.go (hunk 13 above)
Comment (deadprogram) [reply src/machine/machine_rp2040_rom.go:223]: Added range check.

**Review by kenbell — APPROVED** [review]
lgtm - super excited to see this available

**Comment by deadprogram** [comment]
Thank you @kenbell and @bgould for all your help on this!

**Comment by mytechnotalent** [comment]
Thank you all!!!

**Pull request merged**